src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestTPFatTPFServerWithTranslation.java

```diff
@@ -0,0 +1,236 @@
+			} else if (i instanceof SPARQLUnionPattern) {
+				unionTranslation.addAll(handleUnionPattern(((SPARQLUnionPattern) i), fedAccessMgr));
+			} else if (i instanceof SPARQLGroupPattern) {
+				unionTranslation.addAll(handleGroupPattern(((SPARQLGroupPattern) i), fedAccessMgr));
+			} else if (i instanceof BGP) {
+				unionTranslation.addAll(handleBGP(((BGP) i), fedAccessMgr));
+			} else {
+				throw new FederationAccessException(i.toString(), req, fm);
+			}
+		}
+		return unionTranslation;
+	}
+	
+	protected Set<Triple> handleGroupPattern(final SPARQLGroupPattern gp, final FederationAccessManager fedAccessMgr) throws FederationAccessException{
+		Set<Triple> groupTranslation = null;
+		for(final SPARQLGraphPattern i : gp.getSubPatterns()) {
+			if (i instanceof TriplePattern) {
+				if (groupTranslation == null) {
+					groupTranslation = handleTriplePattern((TriplePattern) i, fedAccessMgr);
+				} else {
+					final Set<Triple> partialTranslation = new HashSet<>();
+					for(final Triple j : handleTriplePattern((TriplePattern) i, fedAccessMgr)) {
+						if (groupTranslation.contains(j)) {
+							partialTranslation.add(j);
+						}
+					}
+					groupTranslation = partialTranslation;
+				}
+			} else if (i instanceof SPARQLUnionPattern) {
+				if (groupTranslation == null) {
+					groupTranslation = handleUnionPattern(((SPARQLUnionPattern) i), fedAccessMgr);
+				} else {
+					final Set<Triple> partialTranslation = new HashSet<>();
+					for(final Triple j : handleUnionPattern(((SPARQLUnionPattern) i), fedAccessMgr)) {
+						if (groupTranslation.contains(j)) {
+							partialTranslation.add(j);
+						}
+					}
+					groupTranslation = partialTranslation;
+				}
+			} else if (i instanceof SPARQLGroupPattern) {
+				if (groupTranslation == null) {
+					groupTranslation = handleGroupPattern(((SPARQLGroupPattern) i), fedAccessMgr);
+				} else {
+					final Set<Triple> partialTranslation = new HashSet<>();
+					for(final Triple j : handleGroupPattern(((SPARQLGroupPattern) i), fedAccessMgr)) {
+						if (groupTranslation.contains(j)) {
+							partialTranslation.add(j);
+						}
+					}
+					groupTranslation = partialTranslation;
+				}
+				
+			} else if (i instanceof BGP) {
+				if (groupTranslation == null) {
+					groupTranslation = handleBGP(((BGP) i), fedAccessMgr);
+				} else {
+					final Set<Triple> partialTranslation = new HashSet<>();
+					for(final Triple j : handleBGP(((BGP) i), fedAccessMgr)) {
+						if (groupTranslation.contains(j)) {
+							partialTranslation.add(j);
+						}
+					}
+					groupTranslation = partialTranslation;
+				}
+			} else {
+				throw new FederationAccessException(i.toString(), req, fm);
+			}
+		}
+		return groupTranslation;
+	}
+	
+	protected Set<Triple> handleBGP(final BGP bgp, final FederationAccessManager fedAccessMgr) throws FederationAccessException{
+		Set<Triple> bgpTranslation = null;
+		for(final TriplePattern i : bgp.getTriplePatterns()) {
```

Comment:
I don't think that this is correct, because you are not performing a join between the results of evaluating each triple pattern. Please use/create a Join executable operator, and feed it with the results of evaluating the translation of each triple pattern in the BGP.

Reply:
The triplepatterns which are part of the result are only added if they where already in the solution. Only the first time a request is performed all the solutions are added to the results. Wouldn't this become a join?

Reply:
I don't think so, because we need that the resulting Triples are solutions for ALL the triple patterns in the BGP, so by adding all the triples matching the first triple pattern, you may be adding extra solutions (which would change, if the order of the triple patterns of the BGP changes).